MPS3 AN547 (initial Cortex-M55 support)

galak opened the pull request.

galak committed:
arm: boards: mps3_an547: Add board support for AN547 on MPS3

Add initial board support for the MPS3 AN547.  The board support is
based on the MPS2+ AN521 board support.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak committed:
arm: mps3_an547: Add SoC support for the AN547 on MPS3

The AN547 is a Soft Macro Model implementation of the SSE-300 subsystem
with Ethos-U55 and Cortex-M55 components targeting the MPS3 board.

The SoC support is based on the AN521 MPS2+ support that already exists
in Zephyr.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak committed:
arch: arm: Add initial support for Cortex-M55 Core

Add initial support for the Cortex-M55 Core which is an implementation
of the Armv8.1-M mainline architecture and includes support for the
M‑profile Vector Extension (MVE).

The support is based on the Cortex-M33 support that already exists in
Zephyr.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak commented:
TODO:
[ ] Get toolchain support fixed up for Cortex-M55 (in progress - https://github.com/zephyrproject-rtos/sdk-ng/pull/322)
[ ] fixup docs for MPS3 AN547 (currently copied from AN521 with minor tweaks).

ioannisg reviewed (commented):
Still need to think about it, but maybe we could define and select a new arm architecture kconfig to signify the 8.1 architecture which has more stuff than the v8m, and some minor differences as well.
Something like ARMV8_1_M_MAINLINE

microbuilder commented:
> Still need to think about it, but maybe we could define and select a new arm architecture kconfig to signify the 8.1 architecture which has more stuff than the v8m, and some minor differences as well.
> Something like ARMV8_1_M_MAINLINE

Was just thinking the same ... v8.1 is a bit different than just v8M.

galak commented:
> > Still need to think about it, but maybe we could define and select a new arm architecture kconfig to signify the 8.1 architecture which has more stuff than the v8m, and some minor differences as well.
> > Something like ARMV8_1_M_MAINLINE
> 
> Was just thinking the same ... v8.1 is a bit different than just v8M.

Added a `config ARMV8_1_M_MAINLINE`

microbuilder commented on boards/arm/mps3_an547/Kconfig.board:
```diff
@@ -0,0 +1,6 @@
+# SPDX-License-Identifier: Apache-2.0
+
+config BOARD_MPS3_AN547
+	bool "Arm Cortex-M55 (Corstone SSE-300 w/Ethos-U55) on MPS3 (AN547)"
```
It might make more sense to refer to this as `Corstone-300` since that's the name that ARM uses?

```suggestion
	bool "Arm Cortex-M55 (Corstone-300, SSE-300 w/Ethos-U55) on MPS3 (AN547)"
```

galak replied:
Done.

d3zd3z reviewed (commented):
I'm ok with this, although I don't know enough about the terminology to give feedback on that part.

ioannisg commented on boards/arm/mps3_an547/mps3_an547.dts:
```diff
@@ -0,0 +1,107 @@
+
+/dts-v1/;
+
+#include <arm/armv8-m.dtsi>
```
Should we then have an armv8_1-m.dtsi header? Makes sense to me since we introduce separate Kconfig options.

galak replied:
Will add various v8.1m dtsi/bindings and such.

galak commented:
@gromero fyi.

microbuilder commented on boards/arm/mps3_an547/doc/index.rst:
```diff
@@ -0,0 +1,191 @@
+===========
+
+The MPS3 AN547 has six UARTs. The Zephyr console output by default, uses
+UART0, which is exposed over the Debug USB interface (J8).
```
Is it worth mentionning the need to update the VCP drivers from FTDI for the quad-device chip to work properly on most OSes?

galak replied:
I'd prefer not to get into such details as over time that will probably not be relevant any more, and is so OS dependent (for example MacOS 11.2 didn't need it, my Fedora Linux system was also ok).

microbuilder commented on cmake/gcc-m-cpu.cmake:
```diff
@@ -24,6 +24,12 @@ if("${ARCH}" STREQUAL "arm")
     else()
       set(GCC_M_CPU cortex-m33+nodsp)
     endif()
+  elseif(CONFIG_CPU_CORTEX_M55)
```
Should this also handle a `+nofpu` option if the design is similar to the dual core AN524, where only core1 has the FPU and DSP options, or is `CONFIG_FPU` not being set sufficient?

galak replied:
I feel like maybe we cross that bridge when we get there?  ie the AN547 is a single core so it supports all the features.  When we start seeing other actual SoCs with M55 and/or another FPGA image we can update this as needed

ioannisg commented on CODEOWNERS:
```diff
@@ -343,6 +343,7 @@
 /dts/riscv/microsemi-miv.dtsi             @galak
 /dts/riscv/rv32m1*                        @MaureenHelm
 /dts/riscv/riscv32-litex-vexriscv.dtsi    @mateusz-holenko @kgugala @pgielda
+/dts/arm/armv*m.dtsi                      @galak
```
Could you add my handle here, too @galak?

ioannisg commented on arch/arm/core/aarch32/cortex_m/Kconfig:
```diff
@@ -213,6 +222,16 @@ config ARMV8_M_MAINLINE
+	  This option signifies the use of an ARMv8.1-M processor
+	  implementation, supporting the Main Extension.
+
+	  ARMv8.1-M Main Extension includes additional features
```
I am fine with this.
In theory, the additional features should be reflected in "selection" statements in this symbol definition. But it's ok that we do not have them yet.

ioannisg commented on arch/arm/core/aarch32/cortex_m/mpu/arm_mpu.c:
```diff
@@ -71,7 +71,8 @@ static inline uint8_t get_num_regions(void)
 #include "arm_mpu_v7_internal.h"
 #elif defined(CONFIG_CPU_CORTEX_M23) || \
-	defined(CONFIG_CPU_CORTEX_M33)
+	defined(CONFIG_CPU_CORTEX_M33) || \
```
That is also fine, but I guess it starts getting awkward; perhaps we should have hidden symbols, like ARM_MPU_V8M, and ARM_MPU_V7M, so we do not need to list all CORTEX_M variants. For future improvement.

ioannisg commented on boards/arm/mps3_an547/doc/index.rst:
```diff
@@ -0,0 +1,191 @@
++===========+============+=====================================+
+| NVIC      | on-chip    | nested vector interrupt controller  |
++-----------+------------+-------------------------------------+
+| SYSTICK   | on-chip    | systick                             |
```
Wonder if here you could list SAU, MPU, etc.

galak replied:
We don't seem to do that elsewhere

ioannisg commented on boards/arm/mps3_an547/mps3_an547.yaml:
```diff
@@ -0,0 +1,23 @@
+supported:
+  - gpio
+testing:
+  default: true
```
not objecting to this, but since we do not have any 8.1 special features, maybe it would be enough, for now, that the an521 gets default: true for testing?

galak replied:
Dropped it for now.

ioannisg reviewed (commented):
nice work @galak , please, seem some minor comments I 've added

ioannisg commented:
@galak could you already add a release notes entry for the initial support of ARMv8.1-m?

galak committed:
doc: release-notes-2.6: Updated for Arm-v8.1m/MPS3-AN547 support

Add note about addition of Arm-v8.1m, Cortex-M55, and MPS3-AN547
support being added.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>
